client/state/routing/middleware.js

```diff
@@ -3,32 +3,33 @@
  */
 import debugFactory from 'debug';
 import page from 'page';
-import store from 'store';
 
 /**
  * Internal dependencies
  */
+import localforage from 'lib/localforage';
 import { ROUTE_SET } from 'state/action-types';
 
 const debug = debugFactory( 'calypso:restore-last-location' );
+const key = 'last_path';
 
 let hasInitialized = false;
 
 export const restoreLastLocation = () => ( next ) => ( action ) => {
 	if ( action.type === ROUTE_SET && action.path ) {
-		const lastPath = store.get( 'last_path' );
-
-		if ( ! hasInitialized && lastPath && lastPath !== '/' && action.path === '/' ) {
-			debug( 'redir to', action.path );
-			page( lastPath );
-		} else {
-			debug( 'saving', action.path );
-			store.set( 'last_path', action.path );
-		}
-
-		if ( ! hasInitialized ) {
-			hasInitialized = true;
-		}
+		localforage.getItem( key ).then( ( lastPath ) => {
```

Comment:
Since this is async, we're calling `next( action )` before we can call `page`. It's somewhat minor, but this currently updates the behavior to show a flash of `/` before we're redirected to another place.